[#73] 찜(Pick) 기능 구현

hanwix2 opened the pull request:
- [x] 찜 하기 
- [x] 찜 삭제 (단일 목록)

## 개요:
1. **찜하기**
    - 기존에 존재하는 공연에 대해 찜 목록에 추가하는 기능
    - 찜 하기는 공연 상세 정보 조회 후 한번에 하나의 공연만 찜 가능
2. **찜 삭제 (단일 목록)**
    - 단일 공연 상세 페이지에서 찜 된 상태를 해지
    - 한번의 요청에 하나의 공연만 찜 삭제
    > 다중 목록 삭제는 '찜 목록 조회' 기능 구현 이후에 진행

*일반(GENERAL) 고객 기능*

<br>

## Progress
1. **찜 하기**
    - DB에 찜(pick) 테이블 생성
    - 찜 추가 기능 로직[**insertPick()**]:
        1. 현재 로그인 된 userSession과 찜 할 공연의 performanceId로 찜 추가요청
        2. **checkValidPerformanceId()** - performanceId가 유효한 지 확인 / PerformanceNotExistsException 예외
        3. **getUserNum()** - userSession에서 userId를 활용하여 userNum(user의 id) 획득
        4. **checkPickDuplicated()** - 중복된 찜 목록이 있는지 확인 / PickAlreadyExistsException 예외
        5. **insertPick()** - pick 테이블에 insert
2. **찜 삭제(단일 목록)**
    - PathVariable로 찜에서 삭제할 공연의 performanceId 전달
    - 현재 로그인 된 유저의 userId를 활용해 userNum(User.id) 획득
    - DB의 pick 테이블에서 userNum과 performanceId에 매칭되는 찜 데이터 삭제

<br>

## Test
> 테스트는 Postman을 통해 진행

1. **찜하기:**
    - 성공 case - *일반 회원으로 로그인 후 유효한 공연 id로 찜 추가*
    - 실패1 - *로그인 하지 않은 상태로 찜 추가 요청*
    - 실패2 - *관리자 회원으로 찜 추가 요청*
    - 실패3 - *등록되지 않은 공연 id로 찜 추가 요청*
    - 실패4 - *이미 찜 목록에 등록된 공연을 중복으로 찜 추가 요청*
2. **찜 삭제(단일)**
    - 성공 case - *일반 회원으로 로그인 후 찜 목록에 등록된 공연 id로 찜 삭제*
    - 실패1 - *로그인 하지 않은 상태로 찜 삭제 요청*
    - 실패2 - *관리자 회원으로 찜 삭제 요청*
    - 무시 - *찜 목록에 없는 공연 삭제(아무 일도 일어나지 않음)*

hanwix2 committed:
[#73] 찜하기 기능 구현

- 개요:
    - 기존에 존재하는 공연에 대해 찜 목록에 추가하는 기능
    - 일반(GENERAL) 고객 기능
    - DB에 찜(pick) 테이블 생성
- 찜 추가 기능 로직[insertPick()]:
    1. 현재 로그인 된 userSession과 찜 할 공연의 performanceId로 찜 추가요청
    2. checkValidPerformanceId() - performanceId가 유효한 지 확인 / PerformanceNotExistsException 예외
    3. getUserNum() - userSession에서 userId를 활용하여 userNum(user의 id) 획득
    4. checkPickDuplicated() - 중복된 찜 목록이 있는지 확인 / PickAlreadyExistsException 예외
    5. insertPick() - pick 테이블에 insert

hanwix2 commented on src/main/java/com/show/showticketingservice/service/PickService.java:
```diff
@@ -0,0 +1,35 @@
+
+        performanceService.checkValidPerformanceId(performanceId);
+
+        int userNum = userService.getUserNum(userId);
```
찜 추가 시 현재 로그인 된 유저의 User.id 가 필요합니다.  (userId와 구분하기 위해 여기선 userNum이라고 표현했습니다.)

하지만 현재 UserSession에 저장된 데이터는 User.userId 입니다. 앞으로의 여러 기능 구현에도 효율적인 데이터 관리를 위해 User.id 가 필요할 듯 한데 

**"UserSession에 User.id를 추가하는 작업을 현재 이슈에서 같이 처리해도 될까요?"**

*(다른 기능들이 연관되어 있고, 기존 userId를 이용한 몇몇 로직을 id를 사용하도록 변경하는 작업도 필요합니다.)*

tlsgkr7416 replied:
userNum을 활용 안하고 userId로 활용해도 괜찮지 않나요? pick 테이블에 User.id를 넣어줘야 돼서 그런 거 같은데 거기서는 DB에서 insert 시킬 때 userId를 통해 select문으로 User.id를 받은 후 바로 User.id를 넣어주는건 어떨까요?

hanwix2 replied:
네 그렇게도 생각을 해봤는데요,

앞으로 결제 등 유저 관련 기능에서 
User.id가 많이 쓰일텐데 
매번 User.userId 를 검색해서 하는 것 보단 세션에 저장해서 바로 가져다 쓰는게 효율적이지 않을까 싶어서요~

tlsgkr7416 replied:
그렇군요. 멘토님 의견을 듣는 게 좋겠네요.

f-lab-dev replied:
차라리 추가하기보다는 기존에 `userId`를 아예 `id`로 바꾸는건 어떨까요?

f-lab-dev replied:
혹은 변경의 범위가 너무 커진다 싶으면 말씀하신대로 추가해주는것도 괜찮아보입니다~

hanwix2 replied:
넵! 그럼 따로 이슈 만들어서 처리하겠습니다!

hanwix2 committed:
[#73] 찜 삭제 (단일) 기능 구현

- 개요:
    - 단일 공연 상세 페이지에서 찜 된 상태를 해지
    - 한번의 요청에 하나의 공연만 찜 삭제
    - 일반 고객 기능
- 로직:
    - PathVariable로 찜에서 삭제할 공연의 performanceId 전달
    - 현재 로그인 된 유저의 userId를 활용해 userNum(User.id) 획득
    - DB의 pick 테이블에서 userNum과 performanceId에 매칭되는 찜 데이터 삭제

hanwix2 committed:
[#90] UserSession에 저장되는 userId 속성 변경

- 회원의 userId 보다 id 값을 사용함으로 회원관련 기능의 효용성 증대
- 기존: (String) User.userId
- 변경: (int) User.id

hanwix2 committed:
[#73] UserSession 속성 변경 브랜치 Merge

- merge 시킨 브랜치: feature/90/jh
- 관련 커밋: [e1e3113] UserSession에 저장되는 userId 속성 변경
- 기타 충돌 되는 코드 수정 및 적용
